Fix #481 Allow `default-language` in `package.yaml`

mpilgrem opened the pull request:
The default remains as `Haskell2010` but this also allows `GHC2021`.

The documentation is also updated, accordingly.

Tested by compiling and using with GHC 9.2.1 on Windows.

sol commented:
@mpilgrem I'll try to look at this some time soon.  For the meantime you can use `verbatim` to override `default-language`.

mpilgrem commented:
Looking at the CI failures, something is going wrong with the conditions. I'll have to revisit the pull request.

mpilgrem commented:
I changed my approach which, I think, fixed the problem with the conditionals. I also modified some of the existing tests to specify the section default language as `Haskell2010` in the 'should be' comparisons, which causes the tests to pass (appropriately, I think).

mpilgrem commented:
I have realised that I should not have used `mempty` to introduce the default value of `GhcLanguage`. I will need to revisit the pull request again.

mpilgrem commented:
To explain further my proposed solution:

In module `Hpack.Config`, I introduce type `GhcLanguage` with constructors `Haskell2010` and `GHC2021`, and make it an instance of `FromValue`. It is exported (as it is needed in module `Hpack.Render`). `defaultLanguage` is equated with `Haskell2010`.

I add field `commonOptionsDefaultLanguage :: Last GhcLanguage` to constructor `CommonOptions` (importing `Data.Monoid (Last (..))`).

`defaultCommonOptions` is equated with `mempty` but with `commonOptionsDefaultLanguage = Last (Just defaultLanguage)`.

In `readPackageConfig`, the use of `config` is replaced by `configWithDefaultCommonOptions`, which is `config` updated for `defaultCommonOptions` (using `<>`, hence the use of `Data.Monoid.Last`).

I add field `sectionDefaultLanguage :: Maybe GhcLanguage` to constructor `Section`, and `toSection` equates `sectionDefaultLanguage = getLast commonOptionsDefaultLanguage`.

In module `Data.Aeson.Config.FromValue`, `S1 sel (Rec0 (Last a))` is added as an instance of `GenericDecode`, following the existing instance for `S1 sel (Rec0 (Maybe a))`. A helper operator `.*` is defined to assist with the new instance.

In module `Hpack.Render`, `defaultLanguage` is no longer in the `extraFields` lists and `defaultLanguage` is no longer required.

`renderDefaultLanguage :: GhcLanguage -> Element` is added.

`renderSection` adds `++ maybe [] (return . renderDefaultLanguage) sectionDefaultLanguage` to the end.

andreasabel commented on src/Hpack/Config.hs:
```diff
@@ -501,6 +509,20 @@ instance Semigroup Empty where
   fromValue _ = return Empty
 
+data GhcLanguage =
+    Haskell2010
```
Why not also allow `Haskell98`?

mpilgrem replied:
I don't think there is any reason to exclude Haskell98. I only left it out because the existing `hpack` does not accommodate it.

andreasabel replied:
I think it should be added.  Otherwise the `verbatim: default-language: Haskell98` hack will prevail.

mpilgrem replied:
Now added.

andreasabel reviewed (commented):
This feature would be very much appreciated!!

sol commented:
A better workaround is to use `default-extensions`:

```yaml
default-extensions: GHC2021
```
You will still have `default-language: Haskell2010` in the generated `.cabal` file.  But `GHC2021` will take precedence.

sol commented on src/Data/Aeson/Config/FromValue.hs:
```diff
@@ -143,6 +147,9 @@ instance (Selector sel, FromValue a) => GenericDecode (S1 sel (Rec0 a)) where
   genericDecode = accessFieldWith (.:?)
 
+instance {-# OVERLAPPING #-} (Selector sel, FromValue a) => GenericDecode (S1 sel (Rec0 (Last a))) where
+  genericDecode = accessFieldWith (fmap Last .* (.:?))
```
I think
```haskell
  genericDecode = accessFieldWith (\ value key -> Last <$> (value .:? key))
```
is easier to understand.

I like the use of `Last`, but if you want to add the instance you also need to add tests (similar to what I just did for `Maybe` in https://github.com/sol/hpack/commit/35902395c0dbe2a96b678eef2f93168cd39b80d2, don't add the nested instance, unless we are going to use it though).

Alternatively use `<|>` (as we currently do in other places).

mpilgrem replied:
Noted. I will also retain `Last` and add tests.

sol commented on src/Data/Aeson/Config/FromValue.hs:
```diff
@@ -67,6 +68,9 @@ decodeValue = runParser fromValue
 (.:?) = explicitParseFieldMaybe fromValue
 
+(.*) :: (c -> d) -> (a -> b -> c) -> (a -> b -> d)
+(f .* g) x y = f (g x y)
```
:-1:

sol commented on src/Hpack/Config.hs:
```diff
@@ -185,7 +187,7 @@ packageDependencies Package{..} = nub . sortBy (comparing (lexicographically . f
 
 section :: a -> Section a
-section a = Section a [] mempty [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] Nothing [] mempty mempty []
+section a = Section a [] mempty [] [] [] Nothing [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] [] Nothing [] mempty mempty []
```
👍

sol commented on src/Hpack/Config.hs:
```diff
@@ -501,6 +509,21 @@ instance Semigroup Empty where
+    Haskell98
+  | Haskell2010
+  | GHC2021
+  deriving (Eq, Show, Generic, Enum, Bounded)
```
What about using a `String` instead and leaving anything further to `Cabal`?   That way older versions of `hpack` will continue to work with the next set of `GHC202x` extensions.

A `newtype` will still be neat.  Could we call it `Language` instead of `GhcLanguage`, or is that already taken?

mpilgrem replied:
I will change it to a `newtype` of a `String`, and name the type/constructor `Language`.

sol commented on src/Hpack/Config.hs:
```diff
@@ -501,6 +509,21 @@ instance Semigroup Empty where
+      err = fail ("expected one of " ++ formatOrList ghcLanguagesAsString)
+      ghcLanguages = [minBound .. maxBound]
+      ghcLanguagesAsString = map show ghcLanguages
+      options = zip ghcLanguagesAsString ghcLanguages
```
👎 copy & paste (if we were to keep it, it needs to get DRYed; but see above)

mpilgrem replied:
This instance definition now falls away, due to the change in the type.

sol commented on src/Hpack/Config.hs:
```diff
@@ -652,9 +687,13 @@ readPackageConfig (DecodeOptions programName file mUserDataDir readValue) = runE
+  let Product (Product _ configCommonOptions) _ = config
+      commonOptionsWithDefaults = defaultCommonOptions <> configCommonOptions
+      configWithDefaultCommonOptions =
+        first (second (const commonOptionsWithDefaults)) config
```
Thus feels a little bit dirty, but I don't have any better idea. I think making it more explicit could help (name it clearly after what's going on here, e.g. `configWithDefaultLanguage` etc, don't abstract early).

sol commented on src/Hpack/Render.hs:
```diff
@@ -186,7 +186,7 @@ renderBenchmark (name, sect) =
 
 renderExecutableSection :: [Element] -> Section Executable -> [Element]
-renderExecutableSection extraFields = renderSection renderExecutableFields extraFields [defaultLanguage]
+renderExecutableSection extraFields = renderSection renderExecutableFields extraFields []
```
I think it's now possible to simplify `renderSection` (remove the last argument). Or isn't it?

mpilgrem replied:
It is possible to simplify, and I will do that.

sol commented on src/Hpack/Render.hs:
```diff
@@ -252,6 +252,7 @@ renderSection renderSectionData extraFieldsStart extraFieldsEnd Section{..} = ad
   ++ maybe [] (return . renderBuildable) sectionBuildable
   ++ map (renderConditional renderSectionData) sectionConditionals
   ++ extraFieldsEnd
+  ++ maybe [] (return . renderDefaultLanguage) sectionDefaultLanguage
```
Putting this at the end will prevent you from overriding it in a conditional.  Conditionals have to go last.

sol commented on test/Hpack/ConfigSpec.hs:
```diff
@@ -371,9 +371,9 @@ spec = do
-        , packageTests = Map.fromList [("spec", (section $ executable "Spec.hs") {sectionGhcjsOptions = ["-dedupe", "-ghcjs3"]})]
+          packageLibrary = Just (section library) {sectionGhcjsOptions = ["-dedupe", "-ghcjs1"], sectionDefaultLanguage = Just Haskell2010}
+        , packageExecutables = Map.fromList [("foo", (section $ executable "Main.hs") {sectionGhcjsOptions = ["-dedupe", "-ghcjs2"], sectionDefaultLanguage = Just Haskell2010})]
+        , packageTests = Map.fromList [("spec", (section $ executable "Spec.hs") {sectionGhcjsOptions = ["-dedupe", "-ghcjs3"], sectionDefaultLanguage = Just Haskell2010})]
```
Is it possible to DRY this up?

mpilgrem replied:
It is and I will do so.

sol reviewed (changes requested):
Thanks for opening this PR and sorry for the late response.

This mostly looks great.

I left a couple of comments.

Please add acceptance tests to `test/EndToEndSpec.hs`.

sol commented:
@mpilgrem please don't take any of my comments the wrong way.  It's purely technical.  In general, this project follows TDD.  If you are not familiar with TDD then please let me know and I'll help with the tests.

mpilgrem commented:
@sol, thank you for the review. I think I have made changes to the pull request that are responsive to all of your comments.

sol commented on test/EndToEndSpec.hs:
```diff
@@ -1825,24 +1861,38 @@ library #{name}
 |]
 
+-- As for executable_, but without the final "  default-language: Haskell2020"
+executable'_ :: String -> String -> Package
```
This is a difficult decision. On the one hand we want to keep things DRY (like you did 👍), on the other hand we don't want too many versions of ` executable*` and `library*`. We can't have both.

In this particular case I'm in favor of the duplication. I'll take care of it.

sol reviewed (approved):
This looks great now 👍

I left a couple of notes. I'll merge and take care of them in a follow up.

sol commented:
https://github.com/sol/hpack/pull/501
